src/helpers/messaging.rs

```diff
@@ -143,46 +121,80 @@ impl<N: Network> Mesh<'_, '_, N> {
     }
 }
 
+#[derive(Clone, Copy, Debug)]
+pub struct GatewayConfig {
+    /// Maximum number of items to keep inside the buffer before flushing it to network.
+    /// Note that this buffer is per channel, so setting it to 10 does not imply that every
+    /// 10 messages sent trigger a network request.
+    pub send_buffer_capacity: u32,
+}
+
 impl<N: Network> Gateway<N> {
-    pub fn new(role: Identity, network: N) -> Self {
+    pub fn new(role: Identity, network: N, config: GatewayConfig) -> Self {
         let (tx, mut receive_rx) = mpsc::channel::<ReceiveRequest>(1);
-        let mut message_stream = network.message_stream();
+        let (envelope_tx, mut envelope_rx) = mpsc::channel::<(ChannelId, MessageEnvelope)>(1);
+        let mut message_stream = network.stream();
+        let mut network_sink = network.sink();
+
+        let control_handle = tokio::spawn(async move {
+            // to make forward progress, we periodically check if the system is stalled
+            // if nothing happens for long period of time, we try to unblock it by flushing
+            // the data that remains inside buffers. Note that the interval picked here is somewhat
+            // random - waiting for too long will result in elevated latencies. On the other hand,
+            // sending buffers that are half-full will lead to underutilizing the network
+            const INTERVAL: Duration = Duration::from_millis(200);
```

Comment:
How likely do you believe this to be?  Maybe instead add a flush instruction for a channel that you can use in those cases where everything is done (but you might not have a full batch to send).

Right now, it looks like you are going to need this for every test execution, which I don't like.

Reply:
yea the current mitigation is to set `batch size = 1` in unit tests (happens in `Default` implementation for test config). If batch size is 1, we don't buffer anything and therefore don't rely on this

Reply:
A batch size of 1 works pretty nicely, yeah.